### PR TITLE
Split state groups into a separate data store

### PR DESCRIPTION
**Pull request by erikjohnston**
~~Based on #6294 and #6295.~~

Commits are independently reviewable, though all of this should just be copy and paste from `main` to `state` data stores (modulo updating things to point to the right data store).

(Split out from #6245.)

### CODE REVIEW

File: synapse/storage/data_stores/state/schema/deltas/32/remove_indices.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/* Copyright 2016 OpenMarket Ltd
```
Comment (richvdh) [inline synapse/storage/data_stores/state/schema/deltas/32/remove_indices.sql]: won't having two delta files with the same name mean that only one gets run? or am I misremembering how this works?

File: synapse/storage/prepare_database.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def prepare_database(db_conn, database_engine, config):
     """
 
     # For now we only have the one datastore.
```
Comment (richvdh) [inline synapse/storage/prepare_database.py]: LIES!

**Comment by richvdh** [comment]
> One thought: I'm already beginning to page out what was the idea behind the architecture and the difference between the different storage layers. It might be worth trying to update the text at https://github.com/matrix-org/synapse/blob/master/synapse/storage/__init__.py#L18.

Scratch this; what is needed is for me to actually read it.

Having done so, I think it could do with some expansion and clarification, but that's probably a separate topic.

File: synapse/storage/data_stores/state/schema/deltas/32/remove_indices.sql (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/data_stores/state/schema/deltas/32/remove_indices.sql]: Oh, you're right. I might add a sanity check to cause synapse to explode if there are multiple deltas with the same name

**Comment by erikjohnston** [comment]
Oh, I've also managed to put the state deltas in the wrong place (`deltas/` vs `delta/`), which I didn't pick up due to the full schema.

**Pull request merged**